serde-generate/src/csharp.rs

```diff
@@ -0,0 +1,1170 @@
+    /// short string `name` if possible.
+    fn quote_qualified_name(&self, name: &str) -> String {
+        let qname = self
+            .generator
+            .external_qualified_names
+            .get(name)
+            .cloned()
+            .unwrap_or_else(|| format!("{}.{}", self.generator.config.module_name, name));
+        let mut path = qname.split('.').collect::<Vec<_>>();
+        if path.len() <= 1 {
+            return qname;
+        }
+        let name = path.pop().unwrap();
+        if self.current_reserved_names.contains_key(name) {
+            return qname;
+        }
+        for (index, element) in path.iter().enumerate() {
+            match self.current_namespace.get(index) {
+                Some(e) if e == element => (),
+                _ => {
+                    return qname;
+                }
+            }
+        }
+        name.to_string()
+    }
+
+    fn output_comment(&mut self, name: &str) -> std::io::Result<()> {
+        let mut path = self.current_namespace.clone();
+        path.push(name.to_string());
+        if let Some(doc) = self.generator.config.comments.get(&path) {
+            let text = textwrap::indent(doc, "/// ").replace("\n\n", "\n///\n");
+            write!(self.out, "{}", text)?;
+        }
+        Ok(())
+    }
+
+    fn output_custom_code(&mut self) -> std::io::Result<()> {
+        if let Some(code) = self
+            .generator
+            .config
+            .custom_code
+            .get(&self.current_namespace)
+        {
+            writeln!(self.out, "\n{}", code)?;
+        }
+        Ok(())
+    }
+
+    fn quote_type(&self, format: &Format) -> String {
+        use Format::*;
+        match format {
+            TypeName(x) => self.quote_qualified_name(x),
+            Unit => "Unit".into(),
+            Bool => "bool".into(),
+            I8 => "sbyte".into(),
+            I16 => "short".into(),
+            I32 => "int".into(),
+            I64 => "long".into(),
+            I128 => "BigInteger".into(),
+            U8 => "byte".into(),
+            U16 => "ushort".into(),
+            U32 => "uint".into(),
+            U64 => "ulong".into(),
+            U128 => "BigInteger".into(),
+            F32 => "float".into(),
+            F64 => "double".into(),
+            Char => "char".into(),
+            Str => "string".into(),
+            Bytes => "ImmutableArray<byte>".into(),
+
+            Option(format) => format!("Option<{}>", self.quote_type(format)),
+            Seq(format) => format!("List<{}>", self.quote_type(format)),
+            Map { key, value } => format!(
+                "Dictionary<{}, {}>",
+                self.quote_type(key),
+                self.quote_type(value)
+            ),
+            Tuple(formats) => format!("({})", self.quote_types(formats)),
+            TupleArray {
+                content,
+                size: _size,
+            } => format!("{}[]", self.quote_type(content),),
```

Comment:
@mattico Wait, how come this passes the self-equality test?

Reply:
Arrays and other such things use `Enumerable.SequenceEqual`:

https://github.com/novifinancial/serde-reflection/blob/349258eb6b0ddedaeec448cd81e9a4676de0dc61/serde-generate/src/csharp.rs#L827-L840

I just went and looked at the SequenceEqual implementation and it doesn't short-circuit based on the array length, which could be a big deal for `Bytes`. Should use `IStructuralEquatable` instead.

Reply:
I'm now realizing that collections which have reference-type elements also need to use the appropriate `IEqualityComparer` for their elements.

Reply:
Indeed, `SequenceEqual` does not compose. Also, `GetHashCode` is guaranteed to follow `Equals` (not `SequenceEqual`).

Reply:
In Java, `java.util.*` containers have a value-based `equals` and `hashCode` already, so only arrays are problematic (and should avoided) really. Hopefully, it's the same in C#?

Reply:
> In Java, java.util.* containers have a value-based equals and hashCode already, so only arrays are problematic (and should avoided) really. Hopefully, it's the same in C#?

Unfortunately, no. No standard container types that I can find implement value-based equality by default. Going to have to generate equality helpers.

Reply:
I'm a bit confused. You were already using `ImmutableArray`. How about the other immutable collections for maps etc? https://docs.microsoft.com/en-us/dotnet/api/system.collections.immutable

Reply:
None of the immutable collections implement value equality either, I was mistaken.

Reply:
Ok so using `IStructuralEquatable` and `ImmutableArray`, I managed to make in-depth equality work on 2D matrices: https://gist.github.com/matbd/79e130b5302a93fc3917ecc339809842

It looks like this approach applies to any combination of tuples and arrays (and value types), both for equality and hashing, hence making it possible for those types to be keys in a dictionary: https://stackoverflow.com/questions/3609823/what-problem-does-istructuralequatable-and-istructuralcomparable-solve

This doesn't quite solve the problem of structural equality for user-defined structs `Foo` that would contain a dictionary. For now, I suppose that you could just return a runtime error "Unimplemented" from the implementation of `Equals` and `GetHashCode` of such `Foo`.

Alternatively, at this point, it might be just as easy to define our own collection classes (`Serde.Array`, `Serde.Map`, etc) by deriving (say) the immutable ones and overriding `Equals` and `HashCode` (+ a number of constructors/builders I guess?)

Reply:
Yeah it does seem easier to make collection wrappers. It should communicate the behavior better as well.